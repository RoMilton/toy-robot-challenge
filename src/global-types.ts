export type Coords = [number, number];

export enum Direction {
  North = "NORTH",
  East = "EAST",
  South = "SOUTH",
  West = "WEST",
}

export enum Error {
  CommandNotFound = "COMMAND_NOT_FOUND",
  InvalidPlaceArguments = "INVALID_PLACE_ARGUMENTS",
  RobotNotPlaced = "ROBOT_NOT_PLACED",
}
