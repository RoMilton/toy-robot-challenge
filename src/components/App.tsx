import React, { useState } from "react";
import Table from "./Table";
import ErrorMessage from "./ErrorMessage";
import { GRID_SIZE } from "../constants";
import { Direction, Coords, Error } from "../global-types";

export type TurnType = Command.Right | Command.Left;

export enum Command {
  Place = "PLACE",
  Move = "MOVE",
  Right = "RIGHT",
  Left = "LEFT",
  Report = "REPORT",
}

const orderedDirs = Object.values(Direction);

export const tryMove = (
  coords: Coords,
  direction: Direction
): [number, number] => {
  const [x, y] = coords;
  const max = GRID_SIZE - 1;
  if (direction === Direction.North && y < max) {
    return [x, y + 1];
  }
  if (direction === Direction.East && x < max) {
    return [x + 1, y];
  }
  if (direction === Direction.South && y > 0) {
    return [x, y - 1];
  }
  if (direction === Direction.West && x > 0) {
    return [x - 1, y];
  }

  return [x, y];
};

export const turn = (direction: Direction, turn: TurnType) => {
  const currentIndex = orderedDirs.indexOf(direction);
  const nextOrPrevIndex =
    turn === Command.Right ? currentIndex + 1 : currentIndex - 1;
  return nextOrPrevIndex < 0
    ? orderedDirs[orderedDirs.length - 1]
    : orderedDirs[nextOrPrevIndex % orderedDirs.length];
};

export const validatePlaceArgs = (placeArgs: string[]): boolean =>
  placeArgs.length === 3 &&
  [placeArgs[0], placeArgs[1]].every((val) => !isNaN(Number(val))) &&
  [placeArgs[0], placeArgs[1]].every((val) => Number(val) < GRID_SIZE) &&
  Object.values(Direction).includes(placeArgs[2] as Direction);

function App() {
  const [commandInput, setCommandInput] = useState<string>("");
  const [coords, setCoords] = useState<Coords | undefined>();
  const [direction, setDirection] = useState<Direction | undefined>();
  const [error, setError] = useState<Error | undefined>();
  const [showReport, setShowReport] = useState<boolean>(false);

  const submit = (event: React.ChangeEvent<HTMLFormElement>) => {
    event.preventDefault();
    setError(undefined);
    const parts = commandInput.toUpperCase().trim().split(/[ ,]+/);
    const [command, ...commandArgs] = parts;
    if (!Object.values(Command).includes(command as Command)) {
      setError(Error.CommandNotFound);
      return;
    }

    if (command !== Command.Place && coords === undefined) {
      setError(Error.RobotNotPlaced);
    } else {
      setShowReport(false);

      switch (command) {
        case Command.Place:
          if (!validatePlaceArgs(commandArgs)) {
            setError(Error.InvalidPlaceArguments);
          } else {
            const [x, y, direction] = commandArgs;
            const newCoords = [x, y].map((val) => Number(val));
            setCoords(newCoords as Coords);
            setDirection(direction as Direction);
          }
          break;
        case Command.Move:
          const newCoords = tryMove(coords as Coords, direction as Direction);
          setCoords(newCoords);
          break;
        case Command.Left:
        case Command.Right:
          const newDirection = turn(
            direction as Direction,
            command as TurnType
          );
          setDirection(newDirection);
          break;
        case Command.Report:
          setShowReport(true);
          break;
        default:
      }
    }
  };

  return (
    <div className="App">
      <h1>Robot Challenge</h1>
      <div className="App-inner">
        <Table coords={coords} direction={direction} />
        <form onSubmit={submit}>
          <input
            className="input"
            aria-label="Type command"
            name="command"
            type="text"
            placeholder="Type command"
            value={commandInput}
            onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
              setCommandInput(event.target.value);
            }}
          />
          <button type="submit">Enter</button>
        </form>
        <div className="report">
          {showReport &&
            coords &&
            `Robot is at [${coords[0]}, ${coords[1]}] & facing ${direction}`}
        </div>
      </div>
      <ErrorMessage error={error} />
      <div className="sample">
        Try commands: <br />
        PLACE 0,0,NORTH <br />
        MOVE <br />
        REPORT <br />
      </div>
    </div>
  );
}

export default App;
