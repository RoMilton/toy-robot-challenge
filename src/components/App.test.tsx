import React from "react";
import { render, screen } from "@testing-library/react";
import App, {
  Command,
  TurnType,
  tryMove,
  turn,
  validatePlaceArgs,
} from "./App";
import { Direction } from "../global-types";

type moveTestTuple = [number, number, Direction, number, number];

const moveCases: moveTestTuple[] = [
  [0, 0, Direction.North, 0, 1],
  [2, 4, Direction.South, 2, 3],
  [2, 0, Direction.South, 2, 0],
  [4, 4, Direction.East, 4, 4],
  [2, 1, Direction.East, 3, 1],
  [2, 3, Direction.West, 1, 3],
  [3, 4, Direction.East, 4, 4],
];

describe("tryMove", () => {
  test.each<moveTestTuple>(moveCases)(
    "given position %i %i and facing %s, should return %i %i",
    (x, y, Direction, newX, newY) => {
      expect(tryMove([x, y], Direction)).toEqual([newX, newY]);
    }
  );
});

type turnTestTuple = [Direction, TurnType, Direction];

const turnCases: turnTestTuple[] = [
  [Direction.North, Command.Right, Direction.East],
  [Direction.South, Command.Right, Direction.West],
  [Direction.West, Command.Right, Direction.North],
  [Direction.North, Command.Left, Direction.West],
  [Direction.South, Command.Left, Direction.East],
  [Direction.East, Command.Left, Direction.North],
];

describe("turn", () => {
  test.each<turnTestTuple>(turnCases)(
    "given %s and turning %s, should return %s",
    (direction, turnType, expected) => {
      expect(turn(direction, turnType)).toEqual(expected);
    }
  );
});

type placeArgsTestTuple = [string, string?, Direction?, boolean?];

const placeArgsCases: placeArgsTestTuple[] = [
  ["0", "10", Direction.North, false],
  ["1", "4", Direction.North, true],
];

describe("validatePlaceArgs", () => {
  test.each<placeArgsTestTuple>(placeArgsCases)(
    "given args %s %s %s should return %s",
    (...vals) => {
      const args = vals.slice(0, -1);
      expect(validatePlaceArgs(args as string[])).toEqual(
        vals[vals.length - 1]
      );
    }
  );
});

test("renders App", () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});
