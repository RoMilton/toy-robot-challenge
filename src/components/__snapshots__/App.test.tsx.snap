// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <h1>
      Robot Challenge
    </h1>
    <div
      class="App-inner"
    >
      <div
        class="table"
      >
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
        <div
          class="table-cell"
          style="flex-basis: 20%;"
        />
      </div>
      <form>
        <input
          aria-label="Type command"
          class="input"
          name="command"
          placeholder="Type command"
          type="text"
          value=""
        />
        <button
          type="submit"
        >
          Enter
        </button>
      </form>
      <div
        class="report"
      />
    </div>
    <div
      class="sample"
    >
      Try commands: 
      <br />
      PLACE 0,0,NORTH 
      <br />
      MOVE 
      <br />
      REPORT 
      <br />
    </div>
  </div>
</DocumentFragment>
`;
